File: brokenimagefix/src/main/resources/com/senderproof/vaadin/addons/brokenimagefix/broken-image-fix.js
/**
 * Licensed under the Apache License,Version2.0(the"License");you may not
 * use this file except in compliance with the License.You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,software
 * distributed under the License is distributed on an"AS IS"BASIS,WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND,either express or implied.See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/**
 * JavaScript class for extending the vaadin functionality. This way it is possible to fix broken images in browsers.
 * It is checked whether the images are present or not. If an image isn't available, a CSS class will be added to the
 * parent element of the image (an image in vaadin is always wrapped by a div container).
 */
window.com_senderproof_vaadin_addons_brokenimagefix_BrokenImageFix = function() {

    //noinspection JSUnusedGlobalSymbols
    /**
     * JavaScript methode, which is called through vaadin
     * @param options name of the container, which holds the possibly broken images.
     */
    this.fixBrokenImages = function(options) {
        // we are using the functionality of a JavaScript library (http://imagesloaded.desandro.com/), to check, whether the images are present or not.
        $('#' + options).imagesLoaded().progress(function(instance, image) {
            if (!image.isLoaded) {
                // Image couldn't be loaded, so we add the CSS class to the parent of the image
                var $item = $( image.img ).parent();
                $item.addClass("is-broken");
            }
        });
    };

};

File: brokenimagefix/src/main/java/com/senderproof/vaadin/addons/brokenimagefix/BrokenImageFix.java
/**
 * Licensed under the Apache License,Version2.0(the"License");you may not
 * use this file except in compliance with the License.You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,software
 * distributed under the License is distributed on an"AS IS"BASIS,WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND,either express or implied.See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.senderproof.vaadin.addons.brokenimagefix;

import com.vaadin.annotations.JavaScript;
import com.vaadin.server.AbstractJavaScriptExtension;
import com.vaadin.ui.AbstractLayout;

/**
 * Fix for broken images in browsers. If you use an {@link com.vaadin.ui.Image} with
 * an {@link com.vaadin.server.ExternalResource} in modern browsers and the image isn't available, a broken image is shown. Since there
 * is no built-in functionality to hide these images, there was the need to develop this feature.<br><br>
 * This JavaScript Extension checks the given container for missing images and hides those. Therefore a style class is added to the
 * surrounding container of the image. By default, the image itself is hidden and a background image is set to the surrounding
 * container. This way the user is able to tell that there is an image, which isn't ready. CSS can be customized.<br><br>
 * "is-broken" is the CSS class, which will be added to the surrounding containers of the missing images.<br><br>
 * The server side part has to extend the container, where the possibly missing images are located.<p>
 * <code>
 * VerticalLayout layout = new VerticalLayout();<br>
 * layout.setId("container");<br>
 * new BrokenImageFix().extend(layout, layout.getId());
 * </code></p>
 * This JavaScript Extension relies on jQuery (<a href="https://jquery.com/">https://jquery.com/</a>) and
 * imagesLoaded (<a href="http://imagesloaded.desandro.com/">http://imagesloaded.desandro.com/</a>).
 *
 * @author Marcel Hartwig, senderproof GmbH
 * @version 0.0.1
 */
@JavaScript({ "jquery_min.js", "broken-image-fix.js", "imagesloaded.js" })
public class BrokenImageFix extends AbstractJavaScriptExtension {

    /**
     * This method extends the layout and calls the client side JavaScript, so that the broken images wil be hidden.
     * @param layout layout layout to be extended
     */
    public void extend(final AbstractLayout layout) {
        super.extend(layout);
        this.fixBrokenImages(layout.getId());
    }

    /**
     * Calls the client side JavaScript in order to fix broken images.
     * @param layoutId ID of the container containing the possibly missing images.
     */
    private void fixBrokenImages(final String layoutId) {
        this.callFunction("fixBrokenImages", layoutId);
    }

}
